cds/webapi/C#/BlazorCDS/BlazorCDS/Pages/EnvironmentChooser.razor

```diff
@@ -0,0 +1,178 @@
+@inject IHttpClientFactory ClientFactory
+@inject ILogger<EnvironmentChooser> logger
+@inject Models.AppState AppState
+
+<AuthorizeView>
+    @*Only show the list if the user is signed in and authorized*@
+    <Authorized>
+        @if (instances != null)
+        {
+            <div>
+                <label for="Environments">Choose an Environment:</label>
+                <select name="Environments" id="EnvironmentsSelector" @bind="@SelectedEnvUrl">
+                    @{ int i = 1;
+                        foreach (Instance env in instances.value)
+                        {
+                            if (i == 1)
+                            {
+                                <option selected="selected" value="@env.ApiUrl">@env.FriendlyName</option>
+                                i++;
+                            }
+                            else
+                            {
+                                <option value="@env.ApiUrl">@env.FriendlyName</option>
+                            }
+                        }
+                    }
+                </select>
+            </div>
+        }
+        else
+        {
+            <label>Loading Environments...</label>
+        }
+    </Authorized>
+</AuthorizeView>
+
+
+@code {
+    private string selectedEnvUrl { get; set; }
+
+    [Parameter]
+    public string SelectedEnvUrl
+    {
+        get
+        {
+            return selectedEnvUrl;
+        }
+        set
+        {
+            if (selectedEnvUrl != value)
+            {
+                selectedEnvUrl = value;
+                AppState.SetEnvUrl(value);
+
+                if (SelectedEnvUrlChanged.HasDelegate)
+                {
+                    SelectedEnvUrlChanged.InvokeAsync(value);
+                }
+            }
+        }
+    }
+
+    [Parameter]
+    public EventCallback<string> SelectedEnvUrlChanged { get; set; }
+
+    private InstanceCollection instances;
+
+    protected override async Task OnInitializedAsync()
+    {
+        try
+        {
+
+            var tokenResult = await TokenProvider.RequestAccessToken();
+
+            logger.LogInformation($"AccessTokenResult.Status: {tokenResult.Status}"); //RequiresRedirect
+            logger.LogInformation($"AccessTokenResult.RedirectUrl: {tokenResult.RedirectUrl}");
+
+            if (tokenResult.TryGetToken(out var token))
+            {
+                logger.LogInformation(token.ToString());
+            }
+            else
+            {
+                logger.LogInformation("No accesstoken");
+            }
+
+            //Creates an HttpClient based on the named definition found in Program.Main
+            var client = ClientFactory.CreateClient("GDSClient");
+
+            //Prepare the request to get the data
+            var request = new HttpRequestMessage()
+            {
+                Method = HttpMethod.Get,
+                RequestUri = new Uri($"{client.BaseAddress}api/discovery/v2.0/Instances")
```

Comment:
I'm concerned that this is linked to a `{client.BaseAddress}` variable.
We know that this is `https://globaldisco.crm.dynamics.com/` and it is defined in the client factory in Program.Main.
Hard coding it here would violate the principle of defining this in one place, but leaving this as a variable suggests it might change.

According to [HttpRequestMessage.RequestUri Property](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httprequestmessage.requesturi?view=netstandard-2.1)

> If the request Uri is a relative Uri, it will be combined with the BaseAddress.

Can we just apply a relative Uri here and let it be combined with the base address automatically without using this variable?

Reply:
agreed. removed client.BaseAddress and made it work with relative Uri